src/Compilers/CSharp/Test/Emit/CodeGen/CodeGenStackAllocInitializerTests.cs

```diff
@@ -949,6 +998,340 @@ .maxstack  4
 ");
         }
 
+        [Fact, WorkItem("https://github.com/dotnet/roslyn/issues/69325")]
+        public void TestReadOnlySpan_Net7()
+        {
+            var source = """
+                using System;
+                static class C
+                {
+                    static void Main()
+                    {
+                        ReadOnlySpan<int> p = stackalloc int[3] { 1, 2, 3 };
+                        Write(p);
+                    }
+
+                    static void Write(ReadOnlySpan<int> span)
+                    {
+                        foreach (int x in span)
+                            Console.Write(x);
+                    }
+                }
+                """;
+            var verifier = CompileAndVerify(source, expectedOutput: ExecutionConditionUtil.IsCoreClr ? "123" : null,
+                verify: Verification.FailsPEVerify, targetFramework: TargetFramework.Net70);
+            verifier.VerifyDiagnostics();
+            verifier.VerifyIL("C.Main", """
+                {
+                  // Code size       16 (0x10)
+                  .maxstack  1
+                  IL_0000:  ldtoken    "<PrivateImplementationDetails>.__StaticArrayInitTypeSize=12 <PrivateImplementationDetails>.4636993D3E1DA4E9D6B8F87B79E8F7C6D018580D52661950EABC3845C5897A4D"
+                  IL_0005:  call       "System.ReadOnlySpan<int> System.Runtime.CompilerServices.RuntimeHelpers.CreateSpan<int>(System.RuntimeFieldHandle)"
```

Comment:
>IL_0005:  call       "System.ReadOnlySpan<int> System.Runtime.CompilerServices.RuntimeHelpers.CreateSpan<int>(System.RuntimeFieldHandle)" [](http://example.com/codeflow?start=18&length=137)

This looks like a violation of the language specification ([https://github.com/dotnet/csharpstandard/blob/draft-v9/standard/expressions.md#12821-stack-allocation](https://github.com/dotnet/csharpstandard/blob/draft-v9/standard/expressions.md#12821-stack-allocation)):

> A stack allocation expression allocates a block of memory **from the execution stack**

One might actually depend on the fact, and try to access memory directly with that in mind.

Reply:
Wouldn't it be kinda similar to how the spec is technically also violated already when using array creation expressions to create arrays directly assigned to `ReadOnlySpan<T>` instances, which is lowered as just an RVA field? Ie. the spec ([§ 12.8.16.1](12.8.16.1)) says:
> "Array creation expressions are used to create **new instances of array types**. [...] 
Yet this code:
```csharp
ReadOnlySpan<byte> span = new byte[] { 1, 2, 3 };
```
Doesn't actually create any array. Wouldn't this optimization be analogous to that? If someone was trying to write to those span elements under the assumption that it was stack space, they'd already be deliberately breaking memory safety and getting into undefined behavior, because the local is of `ReadOnlySpan<T>` type 🤔

Reply:
> One might actually depend on the fact, and try to access memory directly with that in mind.

It doesn't seem specified at which exact point the stack memory is allocated, i.e., some local could be allocated after or before the array, making direct memory access a bad idea.

Reply:
> some local could be allocated after or before the array

I do not see how this makes any difference. Direct access to the memory doesn't have to involve any other locals.

Reply:
> It doesn't seem specified at which exact point the stack memory is allocated

It actually does: "A stack allocation expression allocates a block of memory **from the execution stack**. The execution stack is **an area of memory where local variables are stored**."

Reply:
> Wouldn't it be kinda similar to how the spec is technically also violated already when using array creation expressions to create arrays directly assigned to `ReadOnlySpan<T>` instances, which is lowered as just an RVA field? ...

Yes it would. And I find it very unfortunate that the spec was violated this way. Sometimes we simply make a mistake, We focus so much on positive aspects, that negative aspects are dismissed or not even given a serious thought. Other times, we might find ourselves in a situation where the current API shape and the current expressiveness of the language simply do not allow a user to use a feature in an efficient way. Changing the shape of the API or making the language more expressive might not be an option at the time, and we have to make a hard decision to bend the rules. This, however, doesn't mean, that from that point on, there are no rules.

Today, the language is more expressive, we added "Collection Expressions" feature. A user can write:
```
System.ReadOnlySpan<int> x = [1,2,3];
```       
This is going to create and initialize the span in the most efficient way, without violating any rules. It is explicitly specified that no assumptions should be made about the specific mechanism used to create the span and where data reside in memory.
Therefore, I do not think there is a need to bend the rules with stack allocation expression, today we have a language construct that allows a user to achieve the goal.